src/CarbonAware.WebApi/test/integrationTests/CarbonAwareControllerTests.cs

```diff
@@ -184,39 +186,143 @@ public async Task EmissionsForecastsBatch_MissingRequiredParams_ReturnsBadReques
+    {
+        var queryStrings = new Dictionary<string, string>();
+        queryStrings[queryString] = value;
 
-        var result = await PostJSONBodyToURI(inputData, batchForecastURI);
+        var endpointURI = ConstructUriWithQueryString(actualHistoricalURI, queryStrings);
+        var result = await _client.GetAsync(endpointURI);
+
+        Assert.That(result, Is.Not.Null);
+        Assert.That(result.StatusCode, Is.EqualTo(HttpStatusCode.BadRequest));
+    }
+
+    [TestCase(false, false, false, TestName = "EmissionsBatchActual Not use location, Not use startTime, Not use endTime expects BadRequest")]
+    [TestCase(true, false, false, TestName = "EmissionsBatchActual Use location, Not use startTime, Not use endTime expects BadRequest")]
+    [TestCase(true, true, false, TestName = "EmissionsBatchActual Use location, Use startTime, Not use endTime expects BadRequest")]
+    public async Task EmissionsBatchActual_MissingRequiredParams_ReturnsBadRequest(bool useLocation, bool useStart, bool useEnd)
+    {
+        if (useLocation && useStart && useEnd)
+        {
+            Assert.Fail("Invalid test");
+        }
+
+        var intensityBatch = new CarbonIntensityBatchDTO();
+        intensityBatch.Location = useLocation ? "eastus" : null;
+        intensityBatch.StartTime = useStart ? DateTimeOffset.Parse("2022-03-01T15:30:00Z") : null;
+        intensityBatch.EndTime = useEnd ? DateTimeOffset.Parse("2022-03-01T18:30:00Z") : null;
        
+        var intesityData = new List<CarbonIntensityBatchDTO>() { intensityBatch };
+
+        var result = await PostJSONBodyToURI(intesityData, batchActualURI);
+
         Assert.That(result, Is.Not.Null);
-        Assert.That(result!.StatusCode, Is.EqualTo(HttpStatusCode.OK));
-        Assert.That(result!.Content, Is.Not.Null);
-        var data = await result!.Content.ReadAsStringAsync();
-        Assert.That(data, Is.Not.Null);
-        var forecasts = JsonSerializer.Deserialize<List<EmissionsForecastDTO>>(data);
-        Assert.That(forecasts, Is.Not.Null);
-        // assert metadata
-        Assert.AreEqual(forecasts!.Count, inputData.Count);
-        Assert.AreEqual(forecasts!.First().Location, inputData.First().Location);
-        Assert.AreEqual(forecasts!.First().DataStartAt, inputData.First().DataStartAt);
-        Assert.AreEqual(forecasts!.First().DataEndAt, inputData.First().DataEndAt);
-        Assert.AreEqual(forecasts!.First().RequestedAt, inputData.First().RequestedAt);
+        Assert.That(result?.StatusCode, Is.EqualTo(HttpStatusCode.BadRequest));
+    }
+
+    [TestCase("2022-1-1T04:05:06Z", "2022-1-2T04:05:06Z", "eastus", 1, TestName = "EmissionsBatchActual expects OK for eastus 1 element")]
+    [TestCase("2021-12-25", "2021-12-26", "westus", 3, TestName = "EmissionsBatchActual expects OK for westus 3 elements")]
+    public async Task EmissionsBatchActual_SupportedDataSources_ReturnsOk(DateTimeOffset start, DateTimeOffset end, string location, int nelems)
```

Comment:
Similar questions to other tests:

- Can you provide details around the value of second test?
- Prefer string dates in the method signature
- Create JSON payloads without knowledge of our DTO classes
- remove content checking except for Is.Not.Null

Reply:
Address all the comments here. One left is the JSON payloads. Seems like this is done by `PostJSONBodyToURI` already.